### PR TITLE
Target allocator Support for the Agent

### PR DESCRIPTION
**Pull request by okankoAMZ**
Title: Merge Collector Improvements and Bug Fixes

Description:
This pull request merges several improvements and bug fixes related to the CloudWatch Agent Collector and Target Allocator components. The changes include:

1. Split Target Allocator to an internal package for better code organization and maintainability.
2. Added collectorId injection to Target Allocator Manager to support multiple collectors.
3. Fixed an issue where a missing marshaler in the Prometheus receiver caused failures during data translation when using YAML tags for Prometheus configuration.
4. Improved the resilience of the Target Allocator Manager by removing the "return failure" on Start of the thread. This ensures that the Target Allocator Manager can keep trying to ping the Target Allocator pod, preventing the loss of Prometheus metrics in case of startup timing issues.

Testing:
- Tested the changes with the corresponding Amazon CloudWatch Agent changes ([PR #1390](https://github.com/aws/amazon-cloudwatch-agent/pull/1390)) and verified that metrics are being published correctly.
- Manually tested the Target Allocator Manager changes and confirmed that it keeps retrying when the initial ping fails, ensuring no loss of metrics.
 
 Related PRs:
 * #239 
 * #240 
 * #241

### CODE REVIEW

File: receiver/prometheusreceiver/targetallocator/config.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+var _ confmap.Unmarshaler = (*PromHTTPSDConfig)(nil)
+
+func (cfg *PromHTTPSDConfig) Unmarshal(componentParser *confmap.Conf) error {
+	cfgMap := componentParser.ToStringMap()
+	if len(cfgMap) == 0 {
+		return nil
+	}
+	cfgMap["url"] = "http://placeholder" // we have to set it as else marshaling will fail
```
Comment (movence) [inline receiver/prometheusreceiver/targetallocator/config.go:74]: should this be reset after unmarshalling?

File: receiver/prometheusreceiver/config_test.go [hunk 2]
```diff
@@ -23,6 +24,20 @@ import (
...
+
+		os.Exit(1)
+	}
+}
+func teardown() {
+	err := os.Unsetenv("POD_NAME")
+	if err != nil {
+		os.Exit(1)
+	}
+}
```
Comment (jefchien) [inline receiver/prometheusreceiver/config_test.go]: nit: Where are these used?

File: receiver/prometheusreceiver/config_test.go [hunk 3]
```diff
@@ -104,6 +119,22 @@ func TestLoadTargetAllocatorConfig(t *testing.T) {
 	assert.Equal(t, 1, len(r2.PrometheusConfig.ScrapeConfigs))
 	assert.Equal(t, "demo", r2.PrometheusConfig.ScrapeConfigs[0].JobName)
 	assert.Equal(t, promModel.Duration(5*time.Second), r2.PrometheusConfig.ScrapeConfigs[0].ScrapeInterval)
+
+	sub, err = cm.Sub(component.NewIDWithName(metadata.Type, "k8Setup").String())
+	require.NoError(t, err)
+
+	os.Setenv("POD_NAME", "collector-1")
```
Comment (jefchien) [inline receiver/prometheusreceiver/config_test.go]: nit: Use `t.Setenv` in tests.

File: receiver/prometheusreceiver/targetallocator/config.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	CollectorID             string                `mapstructure:"collector_id"`
+	HTTPSDConfig            *PromHTTPSDConfig     `mapstructure:"http_sd_config"`
+	HTTPScrapeConfig        *PromHTTPClientConfig `mapstructure:"http_scrape_config"`
+}
+
+var _ confmap.Unmarshaler = (*Config)(nil)
+
+func getPodName() string {
+	return os.Getenv("POD_NAME")
+}
```
Comment (jefchien) [inline receiver/prometheusreceiver/targetallocator/config.go:34]: nit: Seems like an unnecessary function. It's only ever called in one place and doesn't make things clearer.

**Review by sky333999 — DISMISSED** [review]

**Review by jefchien — COMMENTED** [review]
It's failing the linter. Please fix those changes. We want to keep it so most tests are passing.

**Review by jefchien — APPROVED** [review]

**Review by sky333999 — APPROVED** [review]

**Pull request merged**